modules/build/src/main/scala/scala/build/ScalafixArtifacts.scala

```diff
@@ -0,0 +1,157 @@
+import scala.build.internal.Constants
+import scala.build.internal.CsLoggerUtil.*
+
+final case class ScalafixArtifacts(
+  scalafixJars: Seq[os.Path],
+  toolsJars: Seq[os.Path]
+)
+
+object ScalafixArtifacts {
+
+  def artifacts(
+    scalaVersion: String,
+    externalRulesDeps: Seq[Positioned[AnyDependency]],
+    extraRepositories: Seq[Repository],
+    logger: Logger,
+    cache: FileCache[Task]
+  ): Either[BuildException, ScalafixArtifacts] =
+    either {
+      val scalafixProperties =
+        value(fetchOrLoadScalafixProperties(extraRepositories, logger, cache))
+      val key =
+        value(scalafixPropsKey(scalaVersion))
+      val fetchScalaVersion = scalafixProperties.getProperty(key)
+
+      val scalafixDeps =
+        Seq(dep"ch.epfl.scala:scalafix-cli_$fetchScalaVersion:${Constants.scalafixVersion}")
+
+      val scalafix =
+        value(
+          Artifacts.artifacts(
+            scalafixDeps.map(Positioned.none),
+            extraRepositories,
+            None,
+            logger,
+            cache.withMessage(s"Downloading scalafix-cli ${Constants.scalafixVersion}")
+          )
+        )
+
+      val scalaParameters =
+        // Scalafix for scala 3 uses 2.13-published community rules
+        // https://github.com/scalacenter/scalafix/issues/2041
+        if (scalaVersion.startsWith("3")) ScalaParameters(Constants.defaultScala213Version)
+        else ScalaParameters(scalaVersion)
+
+      val tools =
+        value(
+          Artifacts.artifacts(
+            externalRulesDeps,
+            extraRepositories,
+            Some(scalaParameters),
+            logger,
+            cache.withMessage(s"Downloading scalafix.deps")
+          )
+        )
+
+      ScalafixArtifacts(scalafix.map(_._2), tools.map(_._2))
+    }
+
+  private def fetchOrLoadScalafixProperties(
+    extraRepositories: Seq[Repository],
+    logger: Logger,
+    cache: FileCache[Task]
+  ): Either[BuildException, Properties] =
+    either {
+      val cacheDir  = Directories.directories.cacheDir / "scalafix-props-cache"
+      val cachePath = cacheDir / s"scalafix-interfaces-${Constants.scalafixVersion}.properties"
+
+      val content =
+        if (!os.exists(cachePath)) {
+          val interfacesJar = value(fetchScalafixInterfaces(extraRepositories, logger, cache))
+          val propsData     = value(readScalafixProperties(interfacesJar))
+          if (!os.exists(cacheDir)) os.makeDir(cacheDir)
+          os.write(cachePath, propsData)
+          propsData
+        }
+        else os.read(cachePath)
+      val props  = new Properties()
+      val stream = new ByteArrayInputStream(content.getBytes())
+      props.load(stream)
+      props
+    }
+
+  private def fetchScalafixInterfaces(
+    extraRepositories: Seq[Repository],
+    logger: Logger,
+    cache: FileCache[Task]
+  ): Either[BuildException, Path] =
+    either {
+      val scalafixInterfaces = dep"ch.epfl.scala:scalafix-interfaces:${Constants.scalafixVersion}"
```

Comment:
I think we could. I did that to make scalafix upgrades less painful.

Maybe it would be better to move this fetching logic into the build and put values from `scalafix-properties` into `Constants.scalafixScala212` and etc. However, I would turn that into an issue